homeassistant/components/utility_cost/sensor.py

```diff
@@ -0,0 +1,304 @@
+
+        if self._state is None:
+            # This is the first valid update
+            self._reset()
+
+        if old_state is None or old_state.state in invalid_states:
+            # Write state in case we reset or updated the unit of measurement
+            self.async_write_ha_state()
+            return
+
+        if (
+            state_class != SensorStateClass.TOTAL_INCREASING
+            and new_state.attributes.get(ATTR_LAST_RESET)
+            != old_state.attributes.get(ATTR_LAST_RESET)
+        ) or (
+            state_class == SensorStateClass.TOTAL_INCREASING
+            and reset_detected(
+                self.hass,
+                self._utility_sensor_source_id,
+                float(new_state.state),
+                float(old_state.state),
+                new_state,
+            )
+        ):
+            # Utility meter was reset, reset cost sensor too
+            self._reset()
+            self.async_write_ha_state()
+            return
+
+        try:
+            price = Decimal(price_state.state)
+
+            utility_increment = Decimal(new_state.state) - Decimal(old_state.state)
+            cost_increment = utility_increment * price
+
+            self._state += cost_increment
+            _LOGGER.debug("Cost increment success")
+
+        except DecimalException as err:
+            _LOGGER.warning(
+                "Invalid cost update (utility %s to %s, price %s): %s",
+                old_state.state,
+                new_state.state,
+                price_state.state,
+                err,
+            )
+
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        if state := await self.async_get_last_state():
+            try:
+                self._state = Decimal(state.state)
+            except (DecimalException, ValueError) as err:
+                _LOGGER.warning(
+                    "%s could not restore last state %s: %s",
+                    self.entity_id,
+                    state.state,
+                    err,
+                )
+            else:
+                if self._unit_of_measurement is None:
+                    self._unit_of_measurement = state.attributes.get(
+                        ATTR_UNIT_OF_MEASUREMENT
+                    )
+                self._last_period = (
+                    Decimal(state.attributes[ATTR_LAST_PERIOD])
+                    if state.attributes.get(ATTR_LAST_PERIOD)
+                    else Decimal("0")
+                )
+
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass, [self._utility_sensor_source_id], self._update_cost
+            )
+        )
+
+    @property
+    def native_value(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def native_unit_of_measurement(self):
```

Comment:
consider using `self._attr_unit_of_measurement`

Reply:
Can you elaborate, what is the motivation behind this suggestion? 

I tried to replace the property with assignments to `self._attr_unit_of_measurement` whenever our `self._unit_of_measurment` changes, but it doesn't seem to be working. Is there something more to it?

Likewise for the `native_value` and `extra_state_attributes` suggestions below.

Reply:
The base class already handles the property:

https://github.com/home-assistant/core/blob/487782a6d12019b6ffca6a51080fc5b8779f4715/homeassistant/components/sensor/__init__.py#L334

Reply:
Ah, I had to use the spelling `_attr_native_unit_of_measurement` for it to work correctly. However, I do find that the code becomes less clean, since now we have to remember to do two assignments any time the the unit is changed.